File: runtime/go/units/wasi/path_symlink.go
// requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
// Create a symlink at (fd, new_path) whose target is old_path, stored VERBATIM, never pre-resolved to a host path; containment is enforced later, when the link is followed.
// An absolute target cannot be represented within a preopen sandbox, so it is rejected (cap-std does the same; the suite requires
// "/" to fail).
// The link's own parent is resolved NOFOLLOW.
func (w *WASI) wasi_path_symlink(oldPathPtr, oldPathLen, fd, newPathPtr, newPathLen uint32) uint32 {
    target := string(w.memory.read_string(uint64(oldPathPtr), uint64(oldPathLen)))
    if strings.HasPrefix(target, "/") {
        return wasiNotcapable
    }
    newRel := string(w.memory.read_string(uint64(newPathPtr), uint64(newPathLen)))
    linkHost, err := w.resolve_path(fd, newRel, false)
    if err != wasiOk {
        return err
    }
    // Slash-suffixed link name: EEXIST if something is there
    // (non-directories were ENOTDIR in resolve_path), else ENOENT.
    if strings.HasSuffix(newRel, "/") {
        if _, e := os.Lstat(linkHost); e == nil {
            return wasiExist
        }
        return wasiNoent
    }
    if e := os.Symlink(target, linkHost); e != nil {
        return w.fs_errno(e)
    }
    return wasiOk
}

File: runtime/java/units/wasi/path_symlink.java
// requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
// Create a symlink whose content is the guest's target string stored VERBATIM
// (never pre-resolved): confinement happens later, when something follows the
// link (resolve_path's realpath + containment check), not at creation.
// An absolute target is rejected outright — it could only ever escape the
// preopen tree at follow time. The link's own parent is resolved NOFOLLOW.
int wasi_path_symlink(int oldPtr, int oldLen, int fd, int newPtr, int newLen) {
    String target = new String(
        memory.read_string(Integer.toUnsignedLong(oldPtr), Integer.toUnsignedLong(oldLen)),
        java.nio.charset.StandardCharsets.UTF_8);
    String newRel = new String(
        memory.read_string(Integer.toUnsignedLong(newPtr), Integer.toUnsignedLong(newLen)),
        java.nio.charset.StandardCharsets.UTF_8);
    if (target.startsWith("/")) {
        return WASI_NOTCAPABLE;
    }
    Resolved r = resolve_path(fd, newRel, false);
    if (r.errno != WASI_OK) {
        return r.errno;
    }
    java.nio.file.Path linkPath = java.nio.file.Paths.get(r.path);
    // A trailing slash on the link path demands the name resolve to a directory,
    // so a symlink can never be created there: report why precisely (ENOENT if
    // nothing is there, EEXIST if a directory is, ENOTDIR if a plain file is).
    if (newRel.endsWith("/")) {
        if (!java.nio.file.Files.exists(linkPath, java.nio.file.LinkOption.NOFOLLOW_LINKS)) {
            return WASI_NOENT;
        }
        return java.nio.file.Files.isDirectory(linkPath, java.nio.file.LinkOption.NOFOLLOW_LINKS)
            ? WASI_EXIST
            : WASI_NOTDIR;
    }
    try {
        java.nio.file.Files.createSymbolicLink(linkPath, java.nio.file.Paths.get(target));
    } catch (java.io.IOException ex) {
        return fs_errno(ex);
    }
    return WASI_OK;
}
